infra/bisector.py

```diff
@@ -113,39 +112,52 @@ def bisect(commit_old, commit_new, testcase, fuzz_target, build_data):
 
   Returns:
     The commit SHA that introduced the error or None
+
+  Raises:
+    Value Error: when a repo url can't be determine from the project
   """
   local_store_path = tempfile.mkdtemp()
-  repo_url = build_specified_commit.infer_main_repo(build_data.project_name,
-                                                    local_store_path,
-                                                    commit_old)
-  bisect_repo_manager = repo_manager.RepoManager(repo_url, local_store_path)
+  repo_url, repo_name = build_specified_commit.infer_main_repo(
+      build_data.project_name, commit_old)
+  if not repo_url or not repo_name:
+    raise ValueError('Main git repo url can not be determined.')
+
+  bisect_repo_manager = repo_manager.RepoManager(
+      repo_url, local_store_path, repo_name=repo_name)
   commit_list = bisect_repo_manager.get_commit_list(commit_old, commit_new)
   build_specified_commit.build_fuzzer_from_commit(
       build_data.project_name, commit_list[0], bisect_repo_manager.repo_dir,
       build_data.engine, build_data.sanitizer, build_data.architecture,
       bisect_repo_manager)
-  error_code = helper.reproduce_impl(build_data.project_name, fuzz_target,
-                                     False, [], [], testcase)
+  orig_error_code = helper.reproduce_impl(build_data.project_name, fuzz_target,
+                                          False, [], [], testcase)
+
   old_idx = len(commit_list) - 1
   new_idx = 0
-  if len(commit_list) == 1:
-    if not error_code:
-      return None
-    return commit_list[0]
-
-  while old_idx - new_idx != 1:
+  while old_idx - new_idx > 1:
     curr_idx = (old_idx + new_idx) // 2
     build_specified_commit.build_fuzzer_from_commit(
         build_data.project_name, commit_list[curr_idx],
         bisect_repo_manager.repo_dir, build_data.engine, build_data.sanitizer,
         build_data.architecture, bisect_repo_manager)
-    error_exists = (
-        helper.reproduce_impl(build_data.project_name, fuzz_target, False, [],
-                              [], testcase) == error_code)
-    if error_exists == error_code:
+    error_code = helper.reproduce_impl(build_data.project_name, fuzz_target,
+                                       False, [], [], testcase)
+    if orig_error_code == error_code:
       new_idx = curr_idx
     else:
       old_idx = curr_idx
+  if old_idx - new_idx == 1:
```

Comment:
If we reached this state, that would mean `old_idx` and `new_idx` have different behavior (i.e. there is no crash in old, and there is a crash in new; or the opposite, if we do bisection for a fix). That means we should always return `commit_list[new_idx]`, right?

In other words, if the old and the new revisions have different behavior, that means something has changed in the new revision and that is the answer we were looking for.

Reply:
I think lines 149-159 are not needed. Have you tested this, does it make any sense? Might I be missing something?

Reply:
In the event that all of the commits have an error, this method would return the wrong index. It would return one before the oldest commit. But this could be avoided by doing a single check at the beginning.